src/controllers/legacycontrollersettingsfactory.h

```diff
@@ -0,0 +1,77 @@
+#pragma once
+
+#include <QDomElement>
+#include <QLabel>
+#include <QLayout>
+#include <QList>
+#include <QObject>
+#include <QSizePolicy>
+#include <QSpinBox>
+#include <QString>
+#include <QWidget>
+#include <limits>
+#include <memory>
+#include <tuple>
+
+#include "controllers/legacycontrollersettingsfactory.h"
+
+class AbstractLegacyControllerSetting;
+
+/// @brief This class defines an interface that a controller setting type must
+/// implement so it can be used properly by the builder
+/// @tparam T The class implementing this interface
+template<class T>
+class LegacyControllerSettingFactory {
+    inline static LegacyControllerSettingFactory* createFrom(const QDomElement& element) {
+        return new T(element);
+    }
+    inline static bool match(const QDomElement& element) {
+        return T::match(element);
+    }
+};
+
+/// @brief This class is used to dynamically instantiate a controller setting based on its type
+class LegacyControllerSettingBuilder {
+  public:
+    static LegacyControllerSettingBuilder* instance();
+
+    /// @brief Register a new type of setting. This method is used by the
+    /// REGISTER macro, it shouldn't be used directly
+    /// @param match the match function of the new setting
+    /// @param creator the creator function of the new setting
+    /// @return Always true
+    bool registerType(bool (*match)(const QDomElement&),
+            AbstractLegacyControllerSetting* (*creator)(const QDomElement&)) {
+        m_supportedSettings.append(std::make_tuple(match, creator));
+        return true;
+    }
+
+    /// @brief instantiate a new setting from a an XML definition if any valid
+    /// setting was found. The caller is the owner of the instance
+    /// @param element The XML element to parse to build the new setting
+    /// @return an instance if a a supported setting has been found, null
+    /// otherwise
+    static AbstractLegacyControllerSetting* build(const QDomElement& element) {
+        for (const auto& settingType : qAsConst(instance()->m_supportedSettings)) {
+            if (std::get<0>(settingType)(element)) {
+                return std::get<1>(settingType)(element);
+            }
+        }
+
+        return nullptr;
+    }
+
+  private:
+    LegacyControllerSettingBuilder() = default;
+
+    QList<std::tuple<bool (*)(const QDomElement&),
+            AbstractLegacyControllerSetting* (*)(const QDomElement&)>>
+            m_supportedSettings;
+};
+
+#define CONCAT_(x, y) x##y
+#define CONCAT(x, y) CONCAT_(x, y)
+#define REGISTER_LEGACY_CONTROLLER_SETTING(...)                       \
+    bool CONCAT(kSettingRegistered_, __COUNTER__) =                   \
```

Comment:
is kSettingRegistered_ used?

Reply:
No, what matter here is the call to `registerType`, which will register that setting type to the list of know setting on the builder singleton. Since I'm using constants to hook it together at runtime, it must return something and bool seems the best option in that case. I guess I could go for `int` if you would prefer.

Reply:
Ah I understand, this avoids that the call is optimized out. This feels hacky, maybe we can get rid of it .. I will think about it.

Reply:
At this point we initialize the global bool in the public namespace. This forces the that the LegacyControllerIntegerSetting is registered before main(). I think it can be 'extern const bool" (not tested) 

This trick allows us to have a LegacyControllerSettingBuilder that does not know nothing about the possible settings. 
Do we have business requirement for this? 
If not we may consider to move the calls into LegacyControllerSettingBuilder::LegacyControllerSettingBuilder() and make the code more easy to maintain. 
If you prefer to keep this code, add a brief comment.

Reply:
Just found a not portable solution that does not pollute the namespace: 
https://www.tutorialspoint.com/attribute-constructor-and-attribute-destructor-syntaxes-in-c-in-tutorials-point
I wonder if the C++ standard has nothing for it.

Reply:
I guess there is not perfect solutions, but I kinda preferred the option where one doesn't have to maintain a long list of registration and can simply use they macro right after the settings definition (like QMetaObject). I will try using an `extern const bool`

> Do we have business requirement for this?
> If not we may consider to move the calls into LegacyControllerSettingBuilder::LegacyControllerSettingBuilder() 

No requirements as such, but I wanted to achieves a clean separation of concern here, and make the builder agnostic of all the various type that may exist in the future

Reply:
`extern const bool` is leading to linker issue (multiple definition)

Reply:
Oh, that means your trick is not working, because the setting type is registered every time again when this header is included. 
Good to get rid of it.

Reply:
Oh really? Since the call to the macros was in the `cpp`, I didn't think so. It's gone anyway! :)